src/parser.rs

```diff
@@ -607,6 +620,82 @@ impl Parser {
         }
     }
 
+    pub fn record_or_closure(&mut self) -> NodeId {
+        let span_start = self.position();
+        let mut span_end = self.position(); // TODO: make sure we only initialize it expectedly
+
+        let mut is_closure = false;
+        // For the record
+        let mut items = vec![];
+
+        self.lcurly();
+        self.skip_space_and_newlines();
+
+        // Explicit closure case
+        if self.is_pipe() {
+            let args = Some(self.closure_params());
+            let block = self.block(BlockContext::Closure);
+            self.rcurly();
+            span_end = self.position();
+
+            return self.create_node(
+                AstNode::Closure {
+                    params: args,
+                    block,
+                },
+                span_start,
+                span_end,
+            );
+        }
+
+        let rollback_point = self.get_rollback_point();
+        loop {
+            self.skip_space_and_newlines();
+            if self.is_rcurly() {
+                self.rcurly();
+                span_end = self.position();
+                break;
+            }
+            let key = self.simple_expression();
```

Comment:
I would guard these because we'll want to parse commands soon.

Something like:

```rust
let key = if self.is_simple_expression() {
  self.simple_expression()
} else {
  self.error("expected a key value")
};
```

and then later we'll update that to support commands

Reply:
My naive thinking would say that a command is not a simple expressoin and a single name would be a valid key.
And the first key being not a simple expression would switch to the closure mode.

Which makes me realize that I should commit to record after the first successful key value pair.

Reply:
ooooh my. You're right, so we have to start with thinking it's either one, then immediately switch at the colon if it's there. And if it's there, never switch back.

Yeah, after the first successful key/value pair switching permanently to record sounds good.

Reply:
Implemented the commiting to record in https://github.com/jntrnr/new-nu-parser/pull/6/commits/c3e04c51c4c40a0b54bd6ea047f1efc99f4911df